### DISCUSSION TITLE
How to test GUI apps?

### DISCUSSION THREAD

**Discussion by abitrolly** [teemtee/tmt#1218] [Q&A]
There is a [screen recorder](https://packages.fedoraproject.org/pkgs/byzanz/byzanz/) called [`byzanz`] In Fedora. It [doesn't work](https://bugzilla.redhat.com/show_bug.cgi?id=2001192) at least since F34. 

```
✗ byzanz-record            
usage: byzanz-record [OPTIONS] filename
       byzanz-record --help
✗ byzanz-record test.gif
Given area is not inside desktop.
```

When new version of Fedora is released, Bugzilla sends me a reminder that the bug expired. Instead of manually testing the software after each release, I decided to [try `tmt`](https://docs.fedoraproject.org/en-US/ci/tmt/), which is capable to automatically mark packages as broken for release if tests fail.

I created the dir layout with `tmt init`, then created test with `tmt init --template mini`, and then edited the plan install the package and run the same commands I did above.
```
✗ cat plans/basic.fmf 
summary: Basic smoke test for byzanz
prepare:
  - name: install byzanz
    how: install
    package: byzanz
execute:
  script:
    - byzanz-record
    - byzanz-record test.gif
```

`tmt run` correctly selected Fedora 36 for the test run.

```
✗ tmt run                   
/var/tmp/tmt/run-050

/plans/basic
    discover
        how: shell
        summary: 2 tests selected
    provision
        how: virtual
        image: fedora
        user: root
        memory: 2048 MB
        disk: 10 GB
        arch: x86_64
        progress: booting...
        distro: Fedora Linux 36 (Cloud Edition)
        summary: 1 guest provisioned
    prepare
        how: install
        name: install byzanz
        package: byzanz
    
        summary: 1 preparation applied
    execute
        how: tmt
        summary: 2 tests executed 
    report
        how: display
        summary: 2 tests failed
    finish
        guest: stopped
        guest: removed
        summary: 0 tasks completed

total: 2 tests failed
```

The test failed. However, the message in both cases is different than mine.
```
✗ tmt run --last report -vvv
/var/tmp/tmt/run-050

/plans/basic
    report
        how: display
        order: 50
            fail /script-00
                output.txt: /var/tmp/tmt/run-050/plans/basic/execute/data/script-00/output.txt
                content: 
                    Wrong option: Cannot open display: 
                    usage: byzanz-record [OPTIONS] filename
                           byzanz-record --help
            fail /script-01
                output.txt: /var/tmp/tmt/run-050/plans/basic/execute/data/script-01/output.txt
                content: 
                    Wrong option: Cannot open display: 
                    usage: byzanz-record [OPTIONS] filename
                           byzanz-record --help
        summary: 2 tests failed

total: 2 tests failed
```

I guess something is wrong with the testing setup. So how to test the GUI apps?

**Comment by lukaszachy** [comment]
Hi @abitrolly , tmt spawns Cloud Edition which has minimal set of packages installed and doesn't run any graphics.
In the old days of Xorg you could use start Xvfb (X server rendering to memory), set DISPLAY variable and run your app.

In [Beakerlib](https://github.com/beakerlib/beakerlib) are helper functions (you need to install `xorg-x11-server-Xvfb` as well as `beakerlib`). 

I've tried 
```
source /usr/share/beakerlib/beakerlib.sh
rlJournalStart 
rlPhaseStart
rlPhaseStartTest
rlVirtualXStart foo
export DISPLY=$(rlVirtualXGetDisplay foo)
byzanz-record --help
byzanz-record image
```
and it recorded blank screen (as nothing else is running) So I guess  you need to spawn some other gui app, maybe whole desktop manager?
Also nowdays Wayland is a way to go, so byzanz working in Xvfb might have issues in Wayland.

Fedora QA might/should know much better how to test - https://fedoraproject.org/wiki/QA has mailing lists/irc channel contacts.

**Reply by vhumpa** [reply]
Hi, if you need to spawn a test script within a full desktop session remotely, one way is to use 'dogtail-run-headless-next' script from dogtail (python3-dogtail-scripts). It has an option to allow to choose a --session in particular.
But as Lukas has correctly stated, as it looks like that the byzanc is a pure X screen recorder app, the fact that the wayland is the default display solution going a few versions of Fedora back, likely makes it a little less useful and/or may be the cause why the test ain't working. That said wayland can be disabled by placing WaylandEnable=False in the /etc/gdm/custom.conf default section prior starting the gdm service / or any headless script.

**Reply by abitrolly** [reply]
Actually I am interested to repeat `byzanz` behavior from my systems, which uses Wayland. And it looks like `dogtail` doesn't support Wayland too https://gitlab.com/dogtail/dogtail/-/issues/3

**Reply by lukaszachy** [reply]
I found https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1698 which might be promissing. (another option in Xorg days was to spawn tigervnc and export its DISPLAY into your shell running commands, tigervnc doesn't do wayland either https://github.com/TigerVNC/tigervnc/issues/158)

**Reply by lukaszachy** [reply]
What you need to do is:
1. Install some kind of headless graphical server on SUT together with desktop environment (DE)
2. Start server and DE (it might start automatically)
3. Use graphical server envar/socket/d-bus in test so it runs gui commands in that server

Security hardening makes point 3 more difficult.

From TMT PoV: point 1. and setup could be part of prepare step. point 2. either prepare or part of the test itself - depends if you need fresh gui session for each test or not, as well as ways how to propagate connection details to the graphical server from prepare to execute.

**Reply by abitrolly** [reply]
Sound complicated. I'd prefer an example in `tmt` docs that I can just copy paste to test Wayland app.

**Reply by vhumpa** [reply]
> Actually I am interested to repeat `byzanz` behavior from my systems, which uses Wayland. And it looks like `dogtail` doesn't support Wayland too https://gitlab.com/dogtail/dogtail/-/issues/3

Well, yeah, currently wayland-supporting version is not yet merged to master due to some external dep issues.

Anyhoo just the headless script 'dogtail-run-headless-next' doesn't actually use anything from dogtail framework, it simply runs any given script/binary in a full user session under the gdm for convenience, whatever session is default (Xorg or wayland) when run: $ dogtail-run-headless-next ./my_test_script.sh  
Subsequently, whether the script will then use dogtail or any other framework to do anything is it's business.
One can also use it to run say 'bash' to get a command line which is within full graphical session (with all graphical capabilities, polkit running correctly etc. ) while connected remotely via ssh etc.

**Reply by abitrolly** [reply]
Can this script become a part of the official `tmt` toolset? To make test scenarios that are built on top of it more readable.

**Reply by vhumpa** [reply]
I don't know, as I am not associated with TMT. But since that would mean bringing in most of dogtail and its deps just for this purpose, I am going to say it's unlikely, and only test sets requiring that should install it.

**Reply by lukaszachy** [reply]
> Can this script become a part of the official `tmt` toolset? To make test scenarios that are built on top of it more readable.

Interesting idea but it is out of scope what tmt is supposed to be (Test Management Tool). It could be part of some "helper" ansible roles, cookbooks though. However we are still far from having resources to spent on this, even the idea of "helper" roles is still in brainstorming mode.

**Reply by abitrolly** [reply]
@lukaszachy why GUI tests are out of scope for Test Management Tool?

**Reply by lukaszachy** [reply]
Test framework is outside of Test Management Tool - why would tmt bundle/support pytest, cucumber etc? 
There is no single way how to test GUI - it is up to tester to create their tests. tmt will help to run the testsuite according to the plan (which can contain setup but it is up to tester to provide such setup).

**Reply by abitrolly** [reply]
That becomes more and more confusing. I thought that `tmt` is the test framework, with a feature to execute other test frameworks if necessary. If Python is covered by `pytest`, and basic tests are processed by handling output from `bash`, I would expect a cookbook for GUI tests as well. Otherwise I don't see the reason to learn yet another framework.

**Reply by lukaszachy** [reply]
tmt is not a test framework. tmt is to solve 'test management' with 'install & setup setup system so tests can run there' and report results.

**Reply by thrix** [reply]
Yeah, I would also not call it a test framework, it actually is intended to wrap around other test frameworks. We use mainly Beakerlib for shell tests here, for which we have some more fine-grained support, but otherwise it should be happy to host any test framework.
